### ISSUE TITLE
Lora_gateway program running CPU usage reached 100%

### ISSUE THREAD

**Issue by ylzmdmoon** [CongducPham/LowCostLoRaGw#211]
Lora_gateway program running CPU usage reached 100%;
./lora_gateway Starting with a separate gateway or starting with post_processing_gw.py, the CPU usage is almost 100%. I am running on raspiberry 3B+. Is this normal?

**Comment by CongducPham** [comment]
Yes, as the RPI constantly listen for radio packet at both low-level and high-level (post-processing). However, it doesn't mean that you cant' launch other program. We are trying to optimize the way the low-level and the high-level gateway interact each other.

**Comment by ylzmdmoon** [comment]
font{
        line-height: 1.6;
    }
    ul,ol{
        padding-left: 20px;
        list-style-position: inside;
    }


    
thank you for your reply!I use the gps hat as a gateway to lora, and collect gps information to the gateway. The gps hat is installed on the Raspberry Pi. In the case of high CPU usage, the heat is severe. Will this affect the performance of the gateway?


    
        a#ntes-pcmail-signature-default:hover {
            text-decoration: underline;
            color: #199cff;
            cursor: pointer;
        }
        a#ntes-pcmail-signature-default:active {
            text-decoration: underline;
            color: #246fce;
            cursor: pointer;
        }
    

                
                    发自网易邮箱大师                


    On 6/22/2018 14:54，Congduc Pham<notifications@github.com> wrote： 


Yes, as the RPI constantly listen for radio packet at both low-level and high-level (post-processing). However, it doesn't mean that you cant' launch other program. We are trying to optimize the way the low-level and the high-level gateway interact each other.

**Comment by CongducPham** [comment]
We have several gateways running for a long time. Havent' noticed any overheating.
regards,

**Comment by rertini** [comment]
Putting a small delay i.e. delay(1) in the first while loop of function availabledata(...) lowers the CPU usage to 4~5%:

...
        while( (bitRead(value, 4) == 0) && (millis() < exitTime) )
        {
            value = readRegister(REG_IRQ_FLAGS);
// put delay here
delay(1);     
            // Condition to avoid an overflow (DO NOT REMOVE)
 ...

**Comment by CongducPham** [comment]
Thanks for the suggestion!

regards,

**Comment by CongducPham** [comment]
We may provide a compiler option to use DIO interrupt to lower CPU work.

**Comment by rertini** [comment]
Yes, that would be much better.
